File: frontend/src/lib/types/observability.ts
export interface MetricSeries {
  labels: Record<string, string>;
  samples: [number, number | null][];
}
export interface MetricResult {
  status: string;
  series: MetricSeries[];
  sampled_at: string | null;
  reason: string | null;
}
export interface ResourceSummary {
  metrics_complete: boolean;
  sample_coverage_ratio: number | null;
  values: Record<string, number | null>;
  calculated_at: string;
}
export interface RunnerResource {
  container_id: string;
  task_title?: string | null;
  task_key?: string | null;
  profile_name?: string | null;
  harness?: string | null;
  model?: string | null;
  known_cost_usd?: string;
  unknown_cost_runs?: number;
  input_tokens?: number | null;
  output_tokens?: number | null;
  ai_active_seconds?: number | null;
  developer_active_seconds?: number | null;
  wall_seconds?: number;
  resources?: ServiceResource | null;
  runner_run_id: string;
  task_id: string | null;
  team_id: string | null;
  agent_profile_id: string | null;
  container_name: string;
  service_kind: string;
  phase: string | null;
  started_at: string;
  stopped_at: string | null;
  exit_code: number | null;
  oom_killed: boolean | null;
  summary: ResourceSummary | null;
}
export interface ServiceResource {
  memory_peak?: number | null;
  network_rx_rate?: number | null;
  network_tx_rate?: number | null;
  block_read_rate?: number | null;
  block_write_rate?: number | null;
  throttled_rate?: number | null;
  memory_ratio?: number | null;
  memory_failures?: number | null;
  pids?: number | null;
  uptime_seconds?: number | null;
  restart_count?: number | null;
  oom_count?: number | null;
  exit_code?: number | null;
  state?: string;
  health?: string;
  id: string;
  name: string | null;
  service: string | null;
  cpu?: number | null;
  memory?: number | null;
  memory_limit?: number | null;
  last_seen?: number | null;
}
export interface LiveMetrics {
  events?: InfrastructureEvent[];
  status: string;
  sampled_at: string | null;
  data_freshness_seconds: number | null;
  host: Record<string, number | null>;
  services: ServiceResource[];
  active_runners: RunnerResource[];
  availability: MetricResult;
  targets: MetricResult;
  cpu_definition: string;
}
export interface Incident {
  id: string;
  service_key: string;
  kind: string;
  severity: string;
  opened_at: string;
  closed_at: string | null;
  source: string;
  summary: string;
}
export interface MonitoringSettings {
  monthly_budget_usd?: number | null;
  forecast_horizon_days?: number;
  enabled: boolean;
  retention_days: number;
  retention_size: string;
  refresh_seconds: number;
  forecasts_enabled: boolean;
  forecast_min_samples: number;
  cpu_warning: number;
  ram_warning: number;
  disk_warning: number;
  queue_warning_seconds: number;
  alertmanager_enabled: boolean;
  deployment_managed: boolean;
}

export interface InfrastructureEvent {
  id: string;
  occurred_at: string;
  container_id: string;
  service_name: string;
  event_type: string;
  runner_run_id: string | null;
  exit_code: number | null;
}
export interface Availability {
  status: string;
  days: number;
  services: {
    service: string;
    uptime_ratio: number | null;
    sample_coverage_ratio: number;
    valid_samples: number;
    samples: [number, number | null][];
  }[];
  basis: string;
}

File: backend/app/interfaces/http/schemas/observability.py
"""Public response contracts. Domain query ports remain framework-independent."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ExtensibleResponse(BaseModel):
    model_config = ConfigDict(extra="allow")


class MetricSeriesResponse(BaseModel):
    labels: dict[str, str]
    samples: list[tuple[float, float | None]]


class MetricResponse(BaseModel):
    status: str
    series: list[MetricSeriesResponse] = Field(default_factory=list)
    sampled_at: datetime | None = None
    reason: str | None = None


class ServiceResourceResponse(ExtensibleResponse):
    id: str
    name: str | None = None
    service: str | None = None
    cpu: float | None = None
    memory: float | None = None
    memory_limit: float | None = None
    memory_peak: float | None = None
    network_rx_rate: float | None = None
    network_tx_rate: float | None = None
    block_read_rate: float | None = None
    block_write_rate: float | None = None
    restart_count: int | None = None
    exit_code: int | None = None
    uptime_seconds: float | None = None
    last_seen: float | None = None
    health: str | None = None
    state: str | None = None


class ResourceSummaryResponse(ExtensibleResponse):
    metrics_complete: bool
    sample_coverage_ratio: float | None
    values: dict[str, float | None]
    calculated_at: datetime


class RunnerResponse(ExtensibleResponse):
    runner_run_id: UUID
    container_id: str
    container_name: str
    task_id: UUID | None
    team_id: UUID | None
    agent_profile_id: UUID | None
    service_kind: str
    phase: str | None
    started_at: datetime
    stopped_at: datetime | None
    exit_code: int | None
    oom_killed: bool | None
    summary: ResourceSummaryResponse | None
    resources: ServiceResourceResponse | None = None


class EventResponse(ExtensibleResponse):
    id: UUID
    occurred_at: datetime
    container_id: str
    service_name: str
    event_type: str
    runner_run_id: UUID | None
    exit_code: int | None


class LiveResponse(ExtensibleResponse):
    status: str
    sampled_at: datetime | None = None
    data_freshness_seconds: float | None = None
    host: dict[str, float | None]
    services: list[ServiceResourceResponse]
    active_runners: list[RunnerResponse]
    events: list[EventResponse] = Field(default_factory=list)
    availability: MetricResponse | None = None
    targets: MetricResponse | None = None


class TaskResourcesResponse(BaseModel):
    task_id: UUID
    runners: list[RunnerResponse]
    events: list[EventResponse]
    basis: str


class IncidentResponse(ExtensibleResponse):
    id: UUID
    service_key: str
    kind: str
    severity: str
    opened_at: datetime
    closed_at: datetime | None
    source: str
    summary: str


class ProbeAvailabilityResponse(BaseModel):
    service: str
    uptime_ratio: float | None
    sample_coverage_ratio: float | None
    valid_samples: float | None
    samples: list[tuple[float, float | None]]


class AvailabilityResponse(ExtensibleResponse):
    status: str
    days: int
    services: list[ProbeAvailabilityResponse]
    basis: str
